src/containers/tokens/components/bonding-curve-form/index.js

```diff
@@ -0,0 +1,172 @@
+import React, { PureComponent } from 'react'
+import PropTypes from 'prop-types'
+
+import Button from '../../../../components/button'
+import { form } from '../../../../utils/form-generator'
+import { required, number, positiveNumber } from '../../../../utils/validation'
+import { weiBNToDecimalString } from '../../../../utils/number'
+
+const {
+  Form: BuyPNKFromBondingCurveForm,
+  isInvalid: getBuyPNKFromBondingCurveFormIsInvalid,
+  submit: submitBuyPNKFromBondingCurveForm
+} = form('buyPNKFromBondingCurveForm', {
+  header: {
+    type: 'header',
+    props: { title: 'BUY PNK' }
+  },
+  explanation: {
+    type: 'info'
+  },
+  amountOfETH: {
+    type: 'text',
+    validate: [required, number, positiveNumber],
+    props: {
+      type: 'number',
+      className: 'Form-noMargins'
+    }
+  },
+  rate: {
+    type: 'info',
+    props: {
+      className: 'Tokens-form-rate'
+    }
+  }
+})
+
+export {
+  getBuyPNKFromBondingCurveFormIsInvalid,
+  submitBuyPNKFromBondingCurveForm
+}
+
+const {
+  Form: SellPNKToBondingCurveForm,
+  isInvalid: getSellPNKToBondingCurveFormIsInvalid,
+  submit: submitSellPNKToBondingCurveForm
+} = form('sellPNKToBondingCurveForm', {
+  header: {
+    type: 'header',
+    props: { title: 'SELL PNK' }
+  },
+  explanation: {
+    type: 'info'
+  },
+  amountOfPNK: {
+    type: 'text',
+    validate: [required, number, positiveNumber],
+    props: {
+      type: 'number',
+      className: 'Form-noMargins'
+    }
+  },
+  rate: {
+    type: 'info',
+    props: {
+      className: 'Tokens-form-rate'
+    }
+  }
+})
+
+export {
+  getSellPNKToBondingCurveFormIsInvalid,
+  submitSellPNKToBondingCurveForm
+}
+
+class BondingCurveForm extends PureComponent {
+  static propTypes = {
+    handleBuyPNK: PropTypes.func.isRequired,
+    validateBuyPNK: PropTypes.func.isRequired,
+    handleSellPNK: PropTypes.func.isRequired,
+    validateSellPNK: PropTypes.func.isRequired,
+    buyPNKFromBondingCurveFormIsInvalid: PropTypes.bool.isRequired,
+    sellPNKToBondingCurveFormIsInvalid: PropTypes.bool.isRequired,
+    submitBuyPNKFromBondingCurveForm: PropTypes.func.isRequired,
+    submitSellPNKToBondingCurveForm: PropTypes.func.isRequired,
+    viewState: PropTypes.shape({
+      estimatedETH: PropTypes.string.isRequired,
+      estimatedPNK: PropTypes.string.isRequired
+    }).isRequired
+  }
+
+  onBuyPNKFormChange(values, dispatch, _props) {
+    dispatch({
+      type: 'ESTIMATE_PNK_FROM_BONDING_CURVE',
+      payload: { ETH: values.amountOfETH }
+    })
+  }
+
+  onSellPNKFormChange(values, dispatch, _props) {
+    dispatch({
+      type: 'ESTIMATE_ETH_FROM_BONDING_CURVE',
+      payload: { PNK: values.amountOfPNK }
+    })
+  }
+
+  render() {
+    const {
+      handleBuyPNK,
+      validateBuyPNK,
+      handleSellPNK,
+      validateSellPNK,
+      buyPNKFromBondingCurveFormIsInvalid,
```

Comment:
Why are these props if they are created here?

Reply:
Actually those are not the functions created here, they are the functions produced by `connect()` here: https://github.com/yushih/kleros-juror-front/blob/develop/src/containers/tokens/index.js#L349

Reply:
Yeah, but then why not just connect them in the form file. It's a bit confusing to export that combined form with markup when all other form files just export a single form. I would have two form files for buy and sell and connect them in the parent.